source/clear-linux/tutorials/autofdo.rst

```diff
@@ -0,0 +1,320 @@
+for the user: which parts are never executed or which ones are more
+worthwhile to optimize. What if we could have that feedback? Well we can; the
+:abbr:`FDO (Feedback-Directed Optimization)` technology makes this magic
+happen.
+
+Feedback-Directed Optimization
+==============================
+
+Traditional :abbr:`FDO (Feedback-Directed Optimization)` in GCC uses static
+instrumentation to collect edge and value profiles. GCC uses execution
+profiles consisting of basic block and edge frequency counts to guide the
+optimizations for things like instruction scheduling, basic block reordering,
+function splitting, and register allocation. According to Ramasamy, Yuan,
+Chen & Hundt, 2008, the current method of FDO in GCC involves the following
+steps:
+
+#. Build an instrumented version of the program for edge and value profiling,
+   the instrumentation build.
+
+#. Run the instrumented version with representative training data to collect
+   the execution profile. These runs typically incur significant overhead due
+   to the additional instrumentation code executed.
+
+#. Build an optimized version of the program by using the collected execution
+   profile to guide the optimizations, the FDO build.
+
+The instrumentation and FDO builds are tightly coupled. GCC requires both
+builds use the same in-line decisions and similar optimization flags to
+ensure that the :abbr:`CFG (Control-Flow Graph)` instrumented in the
+instrumentation build matches the CFG annotated with the profile data in the
+FDO build.
+
+Apply this method to our example:
+
+#. Create an instrumented binary with -fprofile-generate:
+
+   .. code-block:: console
+
+      # gcc sort.c -o sort_instrumented -fprofile-generate
+
+#. Run the binary in order to generate the profile data file with
+   runtime information:
+
+   .. code-block:: console
+
+      # ./sort_instrumented
+
+      Bubble sorting array of 30000 elements
+
+      3622 ms
+
+#. Re-build the source with the profile data as feedback:
+
+   .. code-block:: console
+
+      # gcc -O3 sort.c -o sort_fdo -fprofile-use=sort.gcda
+
+      Bubble sorting array of 30000 elements
+
+      1448 ms
+
+We can see an additional improvement from `-O3` to FDO: 1500 ms -> 1448 ms or
+3.46%. We performed 1500 experiments with much more complex code and they
+show a gain of almost 9%.
+
+This method shows good application performance gains but, in practice, it is
+not commonly used due to the high runtime overhead of profile collection, the
+tedious dual-compile usage model, and the difficulties of generating a
+representative training data set.
+
+Enter AutoFDO
+=============
+
+To overcome the limitations of the current FDO model, we proposed the use of
+AutoFDO. The AutoFDO tool uses `perf`_ to collect sample profiles. A
+standalone tool converts the :file:`perf.data` file into the `gcov` format.
+See the `tool's source code`_ for details.
+
+This new model skips the instrumentation step. Instead, it uses a sampling-
```

Comment:
I agree, not having the constant comparison allows us to concentrate on explaining the actual feature.